src/ucp/tag/rndv.c

```diff
@@ -847,80 +1001,183 @@ static ucs_status_t ucp_rndv_pipeline(ucp_request_t *sreq, ucp_rndv_rtr_hdr_t *r
     ucp_ep_h mem_type_ep;
     ucp_mem_desc_t *mdesc;
-    ucp_request_t *frag_req;
+    ucp_request_t *freq;
+    ucp_request_t *fsreq;
     ucp_rsc_index_t md_index;
     ucs_status_t status;
     int i, num_frags;
-    size_t frag_size, length;
-    size_t offset;
-
-    ucs_assert(!UCP_MEM_IS_HOST(sreq->send.mem_type));
+    size_t max_frag_size, rndv_size, length;
+    size_t offset, rndv_base_offset;
 
     /* check if lane supports host memory, to stage sends through host memory */
     md_attr = ucp_ep_md_attr(sreq->send.ep, sreq->send.lane);
     if (!(md_attr->cap.reg_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST))) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    /* check if there are rma_bw lanes (for get_zcopy) on memtype endpoint */
-    mem_type_ep       = worker->mem_type_ep[sreq->send.mem_type];
-    mem_type_rma_lane = ucp_ep_config(mem_type_ep)->key.rma_bw_lanes[0];
-    if (mem_type_rma_lane == UCP_NULL_LANE) {
-        return UCS_ERR_UNSUPPORTED;
+    rndv_size          = rndv_rtr_hdr->size;
+    max_frag_size      = context->config.ext.rndv_frag_size;
+    rndv_base_offset   = rndv_rtr_hdr->offset;
+    num_frags          = ucs_div_round_up(rndv_size, max_frag_size);
+
+    /* initialize send req state on first fragment rndv request */
+    if (rndv_base_offset == 0) {
+         ucp_request_send_state_reset(sreq, NULL, UCP_REQUEST_SEND_PROTO_RNDV_PUT);
     }
 
-    md_index  = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
-    frag_size = context->config.ext.rndv_frag_size;
-    num_frags = (sreq->send.length + frag_size - 1) / frag_size;
+    /* internal send request allocated on sender side to handle send fragments for RTR */
+    fsreq = ucp_request_get(worker);
+    if (fsreq == NULL) {
+        ucs_fatal("failed to allocate fragment receive request");
+    }
 
-    sreq->send.state.uct_comp.count    = num_frags;
-    sreq->send.state.dt.offset         = 0;
-    sreq->send.rndv_put.remote_request = rndv_rtr_hdr->rreq_ptr;
-    sreq->send.rndv_put.remote_address = rndv_rtr_hdr->address;
+    ucp_request_send_state_init(fsreq, ucp_dt_make_contig(1), 0);
+    fsreq->send.buffer                  = sreq->send.buffer + rndv_base_offset;
+    fsreq->send.length                  = rndv_size;
+    fsreq->send.ep                      = sreq->send.ep;
+    fsreq->send.lane                    = sreq->send.lane;
+    fsreq->send.rndv_put.rkey           = sreq->send.rndv_put.rkey;
+    fsreq->send.rndv_put.uct_rkey       = sreq->send.rndv_put.uct_rkey;
+    fsreq->send.rndv_put.remote_request = rndv_rtr_hdr->rreq_ptr;
+    fsreq->send.rndv_put.remote_address = rndv_rtr_hdr->address;
+    fsreq->send.rndv_put.sreq           = sreq;
+    fsreq->send.state.dt.offset         = 0;
 
     offset = 0;
     for (i = 0; i < num_frags; i++) {
-        length = (i != (num_frags - 1)) ? frag_size : (sreq->send.length - offset);
+        length  = (i == (num_frags - 1)) ? (rndv_size - offset) : max_frag_size;
```

Comment:
`length  =` -> `length =`
it has excessive space

Reply:
fixed